### PR TITLE
Feature/add email password

### PR DESCRIPTION
**Pull request by guy-har**

### CODE REVIEW

File: pkg/auth/errors.go [hunk 1]
```diff
@@ -15,4 +15,6 @@ var (
 	ErrNoField                 = errors.New("no field tagged in struct")
 	ErrInvalidAccessKeyID      = errors.New("invalid access key ID")
 	ErrInvalidSecretAccessKey  = errors.New("invalid secret access key")
+	ErrInvalidPassword         = errors.New("invalid password")
+	ErrNoPassword              = errors.New("user has no password")
```
Comment (shimi9276) [inline pkg/auth/errors.go:19]: Would suggest something like "No password passed for the user"

File: clients/java/src/test/java/io/lakefs/clients/api/model/UserTest.java [hunk 2]
```diff
@@ -64,4 +64,12 @@ public void friendlyNameTest() {
         // TODO: test friendlyName
     }
 
+    /**
+     * Test the property 'email'
+     */
+    @Test
+    public void emailTest() {
+        // TODO: test email
```
Comment (shimi9276) [inline clients/java/src/test/java/io/lakefs/clients/api/model/UserTest.java:72]: No tests?

**Review by shimi9276 — COMMENTED** [review]
LGTM! Please note though that I don't have much experience in Go or with lakeFS, so take it with a pint of salt.

**Comment by arielshaqed** [comment]
Please rebase on `origin/master` (or merge that branch in): the name of a required test has changed, and GitHub is waiting for the new name.

https://lakefs.slack.com/archives/C01APUNTSGH/p1648120054377969?thread_ts=1648120054.377969&cid=C01APUNTSGH

File: clients/java/src/main/java/io/lakefs/clients/api/model/User.java [hunk 3]
```diff
@@ -111,6 +115,29 @@ public void setFriendlyName(String friendlyName) {
...
+  @javax.annotation.Nullable
+  @ApiModelProperty(value = "")
+
+  public String getEmail() {
+    return email;
+  }
+
+
+  public void setEmail(String email) {
+    this.email = email;
```
Comment (idanovo) [inline clients/java/src/main/java/io/lakefs/clients/api/model/User.java:137]: no verification? or do we do it in a different place?

Comment (guy-har) [reply clients/java/src/main/java/io/lakefs/clients/api/model/User.java:137]: Currently no verification, we should add the verification once we [support sending emails](https://github.com/treeverse/lakeFS/issues/3103).

File: pkg/ddl/000036_users_email_password.down.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+BEGIN;
+DROP INDEX email_unique_idx ;
```
Comment (nopcoder) [inline pkg/ddl/000036_users_email_password.down.sql:2]:
```suggestion
DROP INDEX email_unique_idx;
```

**Review by nopcoder — APPROVED** [review]
lgtm, there is a small logging and code improvement suggested

File: pkg/ddl/000036_users_email_password.up.sql [hunk 5]
```diff
@@ -0,0 +1,9 @@
+BEGIN;
+
+ALTER TABLE auth_users
+    ADD COLUMN IF NOT EXISTS email TEXT UNIQUE,
+    ADD COLUMN IF NOT EXISTS encrypted_password bytea ;
+
+CREATE UNIQUE INDEX email_unique_idx on auth_users (LOWER(email));
```
[the remark this answers is not in the document]
Comment (guy-har) [reply pkg/ddl/000036_users_email_password.up.sql:7]: Yes

File: pkg/auth/authenticator.go [hunk 6]
```diff
@@ -56,6 +56,27 @@ func (ca ChainAuthenticator) AuthenticateUser(ctx context.Context, username, pas
...
 
+type EmailAuthenticator struct {
+	AuthService Service
+}
+
+func NewEmailAuthenticator(service Service) *EmailAuthenticator {
+	return &EmailAuthenticator{AuthService: service}
+}
+
+func (e EmailAuthenticator) AuthenticateUser(ctx context.Context, username, password string) (int, error) {
```
[the remark this answers is not in the document]
Comment (guy-har) [reply pkg/auth/authenticator.go:66]: Done

**Pull request merged**